### ISSUE TITLE
Problems with `channels.layers.get_channel_layer()`

### ISSUE THREAD

**Issue by severindellsperger** [CJWorkbench/channels_rabbitmq#34]
Hi,

I've written an own Django management command, which receives messages from a service and then should broadcast received messages to a Consumer.
Now, I have problems to receive the current channel layer in following block:
```
def broadcast_message(self, channel, method_frame, header_frame, body):
        body = json.loads(body)
        channel_layer = channels.layers.get_channel_layer()
        channel_layer.group_send(
            "changes",
            {
                "type": update.message,
                "body": body,
            },
        )
```

It always raises an Exception:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/channels_rabbitmq/core.py", line 143, in __init__
    asyncio.get_running_loop()
RuntimeError: no running event loop

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/app/backend/manage.py", line 22, in <module>
    main()
  File "/usr/src/app/backend/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 330, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 371, in execute
    output = self.handle(*args, **options)
  File "/usr/src/app/backend/updatetopology/management/commands/listenonupdates.py", line 144, in handle
    self.process_messages()
  File "/usr/src/app/backend/updatetopology/management/commands/listenonupdates.py", line 119, in process_messages
    self.consume_messages()
  File "/usr/src/app/backend/updatetopology/management/commands/listenonupdates.py", line 109, in consume_messages
    self.channel.start_consuming()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 1865, in start_consuming
    self._process_data_events(time_limit=None)
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 2026, in _process_data_events
    self.connection.process_data_events(time_limit=time_limit)
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 833, in process_data_events
    self._dispatch_channel_events()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 567, in _dispatch_channel_events
    impl_channel._get_cookie()._dispatch_events()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 1492, in _dispatch_events
    consumer_info.on_message_callback(self, evt.method,
  File "/usr/src/app/backend/updatetopology/management/commands/listenonupdates.py", line 79, in broadcast_message
    channel_layer = channels.layers.get_channel_layer()
  File "/usr/local/lib/python3.9/site-packages/channels/layers.py", line 362, in get_channel_layer
    return channel_layers[alias]
  File "/usr/local/lib/python3.9/site-packages/channels/layers.py", line 78, in __getitem__
    self.backends[key] = self.make_backend(key)
  File "/usr/local/lib/python3.9/site-packages/channels/layers.py", line 44, in make_backend
    return self._make_backend(name, config)
  File "/usr/local/lib/python3.9/site-packages/channels/layers.py", line 74, in _make_backend
    return backend_class(**config)
  File "/usr/local/lib/python3.9/site-packages/channels_rabbitmq/core.py", line 145, in __init__
    raise RuntimeError(
RuntimeError: Refusing to initialize channel layer without a running event loop.

If you're writing a django-channels Worker, channels_rabbitmq does not support
this. Read the way to write workers in the carehare documentation.

If you're calling `async_to_sync()`, the call must be within code run by
`sync_to_async()`. Django Channels guarantees this for Django views. Elsewhere,
beware: `asgiref` *does* let you to call `async_to_sync()` without
`sync_to_async()`, but `channels_rabbitmq` *doesn't* -- hence this error.
```
I also tried to play around with `sync_to_async` and `async_to_sync` - unfortunately without success...

Can anyone help me to fix this problem?

**Comment by severindellsperger** [comment]
#Update:
If I use the official Redis channel layer it is working.
If I use this package, it is only working in the Django Shell.
```
python manage.py shell
Python 3.9.2 (default, Feb 19 2021, 17:11:58) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import channels.layers
>>> channel_layer = channels.layers.get_channel_layer()
>>> from asgiref.sync import async_to_sync
>>> async_to_sync(channel_layer.send)('changes', {"type": "hello"})
>>> async_to_sync(channel_layer.receive)('changes')
{'type': 'hello'}
```

I really don't know why the exception in the BaseCommand is raised.

**Comment by adamhooper** [comment]
Thank you for asking!

The problem is described in the error message:

```
RuntimeError: Refusing to initialize channel layer without a running event loop.
```

To solve the problem, make `broadcast` async (and `await` the `group_send`). Async methods always have an event loop; so `get_channel_layer()` will succeed.

The error, by the way, helps defend against production-only catastrophes. Channels users often erroneously open up a new Redis connection per message; and Channels itself won't make a peep about this problem. `channels_rabbitmq` emphasizes correctness. This error message sometimes disrupts correct code; but it often disrupts mistakes.

**Comment by severindellsperger** [comment]
Yes and No :)

With following implementation the code is working:
```
    async def send_message(self, body):
        channel_layer = channels.layers.get_channel_layer()
        await channel_layer.group_send(
            channel_layer._queue_name + "!" + "groups",
            {
                "type": "chat.message",
                "text": "Hello there!",
            },
        )

    def broadcast_message(self, channel, method_frame, header_frame, body):
        body = json.loads(body)  
        loop = asyncio.get_event_loop()
        loop.run_until_complete(self.send_message(body))
        self.channel.basic_ack(delivery_tag=method_frame.delivery_tag)
```

Unfortunately, we are not able to write a consumer, which reacts to these messages. Our goal was to use these messages as inter-process communication between microservices. It looks like this implementation is not made for this aspect, isn't it?
I think the biggest problem is to write a consumer, which receives enqueued messages and not from a WebSocket. Do you have any experience with it? Personally, I don't like to create a WebSocket from the management to the Consumer just to exchange data...

**Comment by adamhooper** [comment]
This is a common use case! It's why we expose `carehare_connection`.

Here's how to write a consumer:

```
async def handle_message(message: bytes) -> None:
    # Your code goes here
    # If your code raises an exception, the app will crash here. That's healthy.

async def do_things_until_a_serious_error():
    channel_layer = get_channel_layer()

    carehare_connection = await channel_layer.carehare_connection
    # If we can't connect to RabbitMQ, the app will crash here. That's healthy.

    await carehare_connection.queue_declare("my-tasks")
    with carehare_connection.acking_consumer("my-tasks") as consumer:
        async for message in consumer:
            await handle_message()
    # if there's a RabbitMQ connection error, the app will crash here. That's healthy.
    # RabbitMQ will deliver the message elsewhere.

# ... and if you want all this to be in a Django management command:
class Command(BaseCommand):
    help = "Run worker"

    def handle(self, *args, **options):
        asyncio.run(do_things_until_a_serious_error())
```

Any async methods in the consumer can `await get_channel_layer().group_send(...)` to send to a group, or `await get_channel_layer().send(...)` to send to a single consumer.

To enqueue a task:

```
async def enqueue_task_or_raise_serious_error(task: bytes) -> None:
    channel_layer = get_channel_layer()
    carehare_connection = await channel_layer.carehare_connection

    # This next line doesn't belong here: it belongs in your app's startup....
    await carehare_connection.queue_declare("my-tasks")  # only needed once, on app startup

    await carehare_connection.publish(task, routing_key="my-tasks")
```

**Comment by severindellsperger** [comment]
Thanks, I basically understand the code, but I don't understand how I can implement it into my Django project. Maybe we misunderstand if we talk about consumers, I speak of Django Channels consumers, and you are talking of general RabbitMQ consumers, correct?

So let me explain it further, where I have my problems:
I have a management command, which receives messages from a queue (via pika) and then should forward it to a Django Channels consumer (via channels_rabbitmq), which lies in my Django Project.
The sense behind it is to give Django the chance to update the data and then notify the connected WebSocket clients. (Maybe wrong design?)

So, I have a general misunderstanding on how I can configure the files `asgi.py` of my Django project as well as my `routing.py` and `consumers.py` of my "consumer" app in a way that the consumer can receive the sent messages to this consumer (or a group).

I think the problem to write a Django Channels Consumer is following:
I don't know how to "start listening on messages" in the Django channels consumer because I have no initial connection. For example, in a WebSocket consumer, I can implement the logic on the `connect(self)` method.

============================================================================
Here I have the implementation of a working consumer via WebSocket. Personally, I don't like the implementation because I have to create a WebSocket on localhost just to notify Channels that there are some new data available...

```
# consumerapp/consumers.py
class UpdateTopologyConsumer(JsonWebsocketConsumer):
    def connect(self):
        async_to_sync(self.channel_layer.group_add)("websocket", self.channel_name)
        self.accept()

    def receive_json(self, content):
        # Process content
        self.send_json(content)
        async_to_sync(self.channel_layer.group_send)(
            "websocket",
            {
                "type": "websocket.message",
                "text": content,
            },
        )

    def websocket_message(self, event):
        self.send_json(event["text"])

    def disconnect(self, close_code):
        self.close()
```
To get this working, of course, I had to create the necessary routes:
```
# consumerapp/routing.py
websocket_urlpatterns = [url("ws/", consumers.UpdateTopologyConsumer.as_asgi())]
```

And also, the routing of my project."
```
# project/asgi.py

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "prototype.settings")


application = ProtocolTypeRouter(
    {
        # (http->django views is added by default)
        "websocket": URLRouter(updatetopology.routing.websocket_urlpatterns),
    }
)
```

Again, the solution would be to replace the WebSocket communication with the implementation of this library. Do you think this is possible?

**Comment by adamhooper** [comment]
TL;DR delete Pika from your stack and use my above code instead.

I'll just barf out some clarifications....

**"Consumer" means one thing to Django Channels, and another thing to RabbitMQ** (and the rest of the world). In RabbitMQ parlance, "consumer" means "worker". That's what I meant when I said "here's how to write a consumer" [above](https://github.com/CJWorkbench/channels_rabbitmq/issues/34#issuecomment-796823413). In Django Channels, a consumer is a Websockets connection.

With `channels_rabbitmq`, you can create a RabbitMQ consumer -- akin to a Django-Channels "worker" -- using `async with (await get_channel_layer().carehare_connection).acking_consumer(...) as consumer`. Don't use Websockets for work queues.

**Django Channels confuses Websocket connections with RabbitMQ connections**. `channels_rabbitmq` tries to clarify. (The RabbitMQ connection is always available as `await get_channel_layer().carehare_connection`)

**For Websockets, use "Channels"**. That's `await get_channel_layer().send()` and `await get_channel_layer().group_send()`.

**For work queues, use RabbitMQ** (carehare). That's `await (await get_channel_layer().carehare_connection).publish(...)` and the consumer pattern I showed above.

In an ideal world, work queues wouldn't use `get_channel_layer()` at all. That way, they wouldn't need to depend on Django. But Django Channels is a framework, not a library. It manages the connection itself.

**Comment by severindellsperger** [comment]
Okay, I try to keep this short this time :smile: 
So, if I have understood this correctly: 
1. I can write a worker (RabbitMQ consumer) with carehare to retrieve and handle my messages from the RabbitMQ queue, which were sent from another microservice.
2. After the handle of the message, I ca can handle the messages, and can then send a message to my Django Channels Consumer (Websocket) to notify all connected clients,

**Comment by adamhooper** [comment]
Yes, that's exactly right.

For 1 (worker), use [my pattern above](https://github.com/CJWorkbench/channels_rabbitmq/issues/34#issuecomment-796823413)

For 2 (sending a message to Websockets, from the worker): `await get_channel_layer().group_send('my-group', {'Hello': 'world!'})`. You could put this code in `handle_message()`, for instance.

Please let me know how it goes! I'd love to revamp our documentation to make this more obvious.

**Comment by severindellsperger** [comment]
Yeah, the improvement of the docu would be great.
Anyway, I rewrote my consumer/worker (thanks it is now much shorter!)
However, I have an error with the `carehare_connection.acking_consumer`:
```
python manage.py listenonupdates  -v 3
2021-03-11 21:11:33,035 | INFO | Connect to RabbitMQ and subscribe to queues: ['changes']
Closing consumer
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/usr/src/app/backend/updatetopology/management/commands/listenonupdates.py", line 48, in process_messages
    with carehare_connection.acking_consumer("changes") as consumer:
AttributeError: __enter__

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/channels_rabbitmq/reader.py", line 32, in consume_into_multi_queue_until_connection_close
    body, delivery_tag = await consumer.next_delivery()
  File "/usr/local/lib/python3.9/site-packages/carehare/_consume_channel.py", line 196, in next_delivery
    return await _next_delivery(self._queue, self.closed)
  File "/usr/local/lib/python3.9/site-packages/carehare/_consume_channel.py", line 39, in _next_delivery
    done, pending = await asyncio.wait(
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 413, in wait
    return await _wait(fs, timeout, return_when, loop)
  File "/usr/local/lib/python3.9/asyncio/tasks.py", line 529, in _wait
    await waiter
asyncio.exceptions.CancelledError
Traceback (most recent call last):
  File "/usr/src/app/backend/manage.py", line 22, in <module>
    main()
  File "/usr/src/app/backend/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 330, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 371, in execute
    output = self.handle(*args, **options)
  File "/usr/src/app/backend/updatetopology/management/commands/listenonupdates.py", line 56, in handle
    asyncio.run(self.process_messages())
  File "/usr/local/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/usr/src/app/backend/updatetopology/management/commands/listenonupdates.py", line 48, in process_messages
    with carehare_connection.acking_consumer("changes") as consumer:
AttributeError: __enter__
```

Here my code:
```
    async def handle_message(message: bytes) -> None:
        self.logger.info(message)
        # Here I process the messages...
        await get_channel_layer().group_send("websocket", {"Hello": "world!"})

    async def process_messages(self):
        channel_layer = get_channel_layer()
        carehare_connection = await channel_layer.carehare_connection
        await carehare_connection.queue_declare(
            "changes", durable=True, arguments={"x-queue-type": "quorum"}
        )
        with carehare_connection.acking_consumer("changes") as consumer:
            async for message in consumer:
                await handle_message()

    def handle(self, *args, **options):
        asyncio.run(self.process_messages())

```

**Comment by adamhooper** [comment]
> `with carehare_connection.acking_consumer("changes") as consumer:`

Should be `async with`

**Comment by severindellsperger** [comment]
I have definitely get more into the concurrency topics :laughing: 
However, it now works for me - I get updates from a RabbitMQ queue, process the updates (not implemented yet), and afterward send a message to the WebSocket group (group, where all WebSocket clients are listening).

This project is awesome! Thanks for your help man!

============================================================================
Just in case, anyone is interested in the code.
First, I have written a [Django management command](https://docs.djangoproject.com/en/3.1/howto/custom-management-commands/), to run the worker in a separate process. This could maybe be done in another way but works for me.
In this application, I retrieve messages from a message queue and process them. After the messages are processed, I send the message to my WebSocket consumer in Django Channels (see below).

The logic of this program makes no sense (yet), but should show the implementation and thoughts.

```

class Command(BaseCommand):
    help = "Connects to RabbitMQ Queue(s), listens to messages ,and forwards it to the Django Channels Consumer"

    async def handle_message(self, message: bytes) -> None:
        # Here I process the messages...
        # After processing send a notification to all open WebSockets
        await get_channel_layer().group_send(
            "websocket", {"type": "websocket.message", "message": json.loads(message)}
        )

    async def process_messages(self):
        channel_layer = get_channel_layer()
        carehare_connection = await channel_layer.carehare_connection
        await carehare_connection.queue_declare(
            "changes", durable=True, arguments={"x-queue-type": "quorum"}
        )
        async with carehare_connection.acking_consumer("changes") as consumer:
            async for message in consumer:
                await self.handle_message(message)

    def handle(self, *args, **options):
        self.logger.info(f"Connect to RabbitMQ and subscribe to queues: {self.queues}")
        asyncio.run(self.process_messages())
```

Here my Consumer, which simply handles the messages:
```
class MessageConsumer(JsonWebsocketConsumer):
    def connect(self):
        async_to_sync(self.channel_layer.group_add)("websocket", self.channel_name)
        self.accept()

    def websocket_message(self, message):
        print(message, flush=True)
        self.send_json(message)

    def disconnect(self, close_code):
        self.close()

```
